internal/pkg/skuba/replica/replica.go

```diff
@@ -0,0 +1,278 @@
+	fromSelectors, err := client.AppsV1().Deployments(metav1.NamespaceSystem).List(
+		metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("%s=true", highAvailabilitylabel),
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	node, err := kubernetes.GetAllNodes(client)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Helper{
+		Client:            client,
+		ClusterSize:       len(node.Items),
+		MinSize:           minSize,
+		SelectDeployments: *fromSelectors,
+	}, nil
+}
+
+func (r *Helper) replaceAffinity(remove affinity, create affinity) error {
+	// doRemove will remove only when affinity path exists
+	doRemove := false
+	affinity := r.Deployment.Spec.Template.Spec.Affinity
+	if affinity != nil && affinity.PodAntiAffinity != nil {
+		preferredList := len(r.Deployment.Spec.Template.Spec.Affinity.PodAntiAffinity.PreferredDuringSchedulingIgnoredDuringExecution)
+		requiredList := len(r.Deployment.Spec.Template.Spec.Affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution)
+		switch remove {
+		case preferred:
+			if preferredList != 0 {
+				doRemove = true
+				break
+			}
+			if requiredList != 0 {
+				return nil
+			}
+		case required:
+			if requiredList != 0 {
+				doRemove = true
+				break
+			}
+			if preferredList != 0 {
+				return nil
+			}
+		}
+	}
+	if doRemove {
+		if err := r.removeAffinity(); err != nil {
+			return err
+		}
+	}
+
+	var affinityJSON string
+	switch create {
+	case required:
+		affinityJSON = fmt.Sprintf(patchAffinityRequired, r.Deployment.ObjectMeta.Name)
+	default:
+		affinityJSON = fmt.Sprintf(patchAffinityPreferred, r.Deployment.ObjectMeta.Name)
+	}
+
+	_, err := r.Client.AppsV1().Deployments(metav1.NamespaceSystem).Patch(r.Deployment.ObjectMeta.Name, types.StrategicMergePatchType, []byte(affinityJSON))
+	if err != nil {
+		return err
+	}
+
+	r.isUpdated = true
+	return nil
+}
+
+func (r *Helper) removeAffinity() error {
+	_, err := r.Client.AppsV1().Deployments(metav1.NamespaceSystem).Patch(r.Deployment.ObjectMeta.Name, types.JSONPatchType, []byte(patchAffinityRemove))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *Helper) updateDeploymentReplica(size int) (*appsv1.Deployment, error) {
+	replicaJSON := fmt.Sprintf(patchReplicas, size)
+	return r.Client.AppsV1().Deployments(metav1.NamespaceSystem).Patch(r.Deployment.ObjectMeta.Name, types.StrategicMergePatchType, []byte(replicaJSON))
+}
+
+func (r *Helper) refreshReplicas() error {
+	// At least one node needs to be free of the affinity rules for pods to re-distribute.
+	// Sizing down replicas would make room for new pod resource.
+	// For example, during rolling upgrade if all node already has `required`, scheduler will `Pend`
+	// when tries to create a new deployment with `preferred`. This is due to the conflict to its original rule.
+	// But this ends up with less of replicas. So need the following patch to bring the replicas to the right size.
+	if _, err := r.updateDeploymentReplica(r.ClusterSize - 1); err != nil {
```

Comment:
```suggestion
	if _, err := r.updateDeploymentReplica(r.ReplicaSize - 1); err != nil {
```

Shoud we do `r.ReplicaSize - 1` instead of `r.ClusterSize - 1`? (assume r.ReplicaSize is at least 2)

Reply:
There might be more than 1 replica on single node. So there will be chances of that node is still tide with deployment affinity rule when set `r.ReplicaSize - 1`. 
Set `r.ClusterSize - 1` will ensure there is always one node free of affinity rule.

Reply:
Got your point. `refreshReplicas` has it's usage context. Let's keep it for now and improve later if need. Thanks.